### ISSUE TITLE
Can not run java project when Erlang-plugin installed (IJ 2016-1)

### ISSUE THREAD

**Issue by maxifly** [ignatov/intellij-erlang#727]
IntelliJ IDEA 2016.1
Build #IC-145.258, built on March 17, 2016
JRE: 1.8.0_60-b27 x86
JVM: Java HotSpot(TM) Server VM by Oracle Corporation
## Erlang plugin - 0.7.876

When I delete Erlang plugin? restart Idea and try run java project from Idea - project running normaly.

When I install Erlang plugin, restart Idea and try run java project from IntellijIdea, I have error:

Error:Internal error: (java.lang.UnsupportedClassVersionError) com/intellij/execution/ExecutionException : Unsupported major.minor version 52.0
java.lang.UnsupportedClassVersionError: com/intellij/execution/ExecutionException : Unsupported major.minor version 52.0
    at java.lang.ClassLoader.defineClass1(Native Method)
    at java.lang.ClassLoader.defineClass(ClassLoader.java:800)
    at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
    at java.net.URLClassLoader.defineClass(URLClassLoader.java:449)
    at java.net.URLClassLoader.access$100(URLClassLoader.java:71)
    at java.net.URLClassLoader$1.run(URLClassLoader.java:361)
    at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
    at java.security.AccessController.doPrivileged(Native Method)
    at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:425)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:358)
    at org.intellij.erlang.jps.builder.ErlangBuilderService.createBuilders(ErlangBuilderService.java:38)
    at org.jetbrains.jps.incremental.BuilderRegistry.<init>(BuilderRegistry.java:53)
    at org.jetbrains.jps.incremental.BuilderRegistry.<init>(BuilderRegistry.java:33)
    at org.jetbrains.jps.incremental.BuilderRegistry$Holder.<clinit>(BuilderRegistry.java:36)
    at org.jetbrains.jps.incremental.BuilderRegistry.getInstance(BuilderRegistry.java:43)
    at org.jetbrains.jps.cmdline.BuildRunner.runBuild(BuildRunner.java:132)
    at org.jetbrains.jps.cmdline.BuildSession.runBuild(BuildSession.java:294)
    at org.jetbrains.jps.cmdline.BuildSession.run(BuildSession.java:125)
    at org.jetbrains.jps.cmdline.BuildMain$MyMessageHandler$1.run(BuildMain.java:232)
    at org.jetbrains.jps.service.impl.SharedThreadPoolImpl$1.run(SharedThreadPoolImpl.java:44)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
    at java.util.concurrent.FutureTask.run(FutureTask.java:262)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:745)

**Comment by maxifly** [comment]
If I delete from run configuration (in region "Before lanch") step "Make" then  run project work normaly

**Comment by deadok22** [comment]
I took a look at the source, and it appears that com.intellij.compiler.server.BuildManager#getBuildProcessRuntimeSdk attempts to pick up your project's JDK. What's your project's JDK version, @maxifly?

@ignatov, what's the status of switching IntelliJ to Java8? Are jps builders supposed to be run with Java6 only?

**Comment by ignatov** [comment]
I'm going to ask Nik about 1.8 in JPS builders.

**Comment by deadok22** [comment]
@ignatov bump

**Comment by maxifly** [comment]
@deadok22,  JDK version: jdk1.7.0_60

**Comment by deadok22** [comment]
So it looks like IntelliJ does pick up that JDK and runs compiler process with it. 

We'll have to wait for Nik's response on 1.8 and JPS builders.

Meanwhile, could you try building the project with JDK 1.8, @maxifly?

**Comment by KronicDeth** [comment]
@ignatov, @deadok22 any word from Nik?  I'm having similar issues with IntelliJ Elixir builds.  I have to have both intellij-erlang and intellij-elixir uninstalled if I want to use JDK < 1.8.  My current work-around is to build against JDK 1.6 in a copy of IntelliJ 14.1 as outlined [here](https://github.com/KronicDeth/intellij-elixir/issues/290#issuecomment-222693633).  That version can be used in older IntelliJ safely, but as sooner as there's a JDK 1.8 JPS in the install, I can't build against JDK < 1.8.

**Comment by chashnikov** [comment]
Yes, classes included into classpath of the external build process must be compatible with Java 6.
But accordingly to the original exception you have another problem in Erlang plugin: JPS classes cannot depend on classes from openapi.jar (like com/intellij/execution/ExecutionException), they may depend only on classes from jps-builders.jar, jps-model.jar, util.jar and some others. Unfortunately currently I don't see a simple way to enforce this.

**Comment by deadok22** [comment]
@chashnikov enforcing it is simply not putting ```openapi.jar``` on the classpath - what is the problem?

**Comment by chashnikov** [comment]
@deadok22 I don't know how your project is configured, its iml files aren't available at GitHub. So I suppose *.iml files are automatically generated when you import the project from Gradle, so to exclude openapi.jar from classpath for jps-module we need to support this in intellij-gradle-plugin.

**Comment by deadok22** [comment]
@chashnikov you're right. @zolotov, could you take a look?

@chashnikov shouldn't IntelliJ exclude ```openapi.jar``` JPS builders' classpath, too?

**Comment by zolotov** [comment]
@deadok22 filed an issue https://github.com/JetBrains/gradle-intellij-plugin/issues/106

**Comment by KronicDeth** [comment]
@chashnikov I'm still using .iml files as my usage was adapted from the pre-gradle version of intellij-erlang.  How do I fix this for https://github.com/KronicDeth/intellij-elixir?

**Comment by KronicDeth** [comment]
Both intellij-erlang's [`JpsErlangModuleType`](https://github.com/ignatov/intellij-erlang/blob/b88e142925b4b7c448c1d86d695b02bd45f11d34/jps-plugin/src/org/intellij/erlang/jps/model/JpsErlangModuleType.java) and intellij-elixir's [`JpsElixirModuleType`](https://github.com/KronicDeth/intellij-elixir/blob/b65be01afc01ec51b11bdf3749f5d7e0230344f4/jps-shared/src/org/elixir_lang/jps/model/JpsElixirModuleType.java) classes use `org.jetbrains.jps.model.JpsDummyElement`, which is from `openapi.jar`.  So that seems to imply that those classes need to be implemented differently or not be needed for the jps-builder build.

**Comment by chashnikov** [comment]
@KronicDeth In your module it's probably enough to set 'Language Level' to '6' for 'jps-plugin' and 'jps-shared' modules.

**Comment by chashnikov** [comment]
@KronicDeth @zolotov Hmm, it appears that JPS model API classes are packaged to 'openapi.jar' in IDEA distribution ('jps-model.jar' contains JPS model implementation classes only), so we need to include that JAR to the external build process' classpath. However jps-plugin modules may use only JPS-related classes from openapi.jar, otherwise they will be broken when we extract JPS model API classes to a separate JAR.

**Comment by KronicDeth** [comment]
Is there a lower-level class I can use to replace the usages of `com.intellij.execution.configurations.GeneralCommandLine` as it throws `com.intellij.execution.ExecutionException` and that triggers the error (Unsupported major.minor version 52.0) when targeting 14.1-1.6 from 2016.1-1.8:

|                       | IntelliJ | JDK |
|-----------------------|----------|-----|
| Build                 | 14.1     | 1.6 |
| Original Build Target | 14.1     | 1.6 |
| Install               | 2016.1   | 1.8 |
| Install Build Target  | 14.1     | 1.6 |

I was worried that all of `openapi.jar` was `52.0` for 2016.1, so I used [this](http://www.rgagnon.com/javadetails/java-0544.html) to check the MAJOR.MINOR

```java
import java.io.*;

public class ClassVersionChecker {
    public static void main(String[] args) throws IOException {
        for (int i = 0; i < args.length; i++)
            checkClassVersion(args[i]);
    }

    private static void checkClassVersion(String filename)
        throws IOException
    {
        DataInputStream in = new DataInputStream
         (new FileInputStream(filename));

        int magic = in.readInt();
        if(magic != 0xcafebabe) {
          System.out.println(filename + " is not a valid class!");;
        }
        int minor = in.readUnsignedShort();
        int major = in.readUnsignedShort();
        System.out.println(filename + ": " + major + " . " + minor);
        in.close();
    }
}
```
To confirm that, I unpacked the jar and looked at the MAJOR.MINOR in the `.class`

```sh
> jar xf /Applications/IntelliJ\ IDEA.app/Contents/lib/openapi.jar
> javac ClassVersionChecker.java
> java ClassVersionChecker com/intellij/execution/ExecutionException.class
com/intellij/execution/ExecutionException.class: 52 . 0
> java ClassVersionChecker org/jetbrains/jps/model/JpsDummyElement.class
org/jetbrains/jps/model/JpsDummyElement.class: 50 . 0
```

So, `openapi.jar` has a mix of MAJOR.MINOR classes.  I assume because `org/jetbrains` was built for 1.6 and `com/jetbrains` was built for 1.8, so it is possible, to manually verify if a `.class` will not work with `1.8`, but I haven't found a way to get the Project Settings to reject the build ahead of time yet.

So, this leads back to, what can I use besides `GeneralCommandLine` to call command line tools?  
1. Drop down to `ProcessBuilder`, but lose the cross-platform support in `GeneralCommandLine`
2. Copy `GeneralCommandLine` into my jps-builder just so I can build it in my own namespace with 1.6.

Any guidance what's the better choice here?

**Comment by KronicDeth** [comment]
I went through checking the MAJOR.MINOR of all the `imports` in `ErlangBuilder` and `ElixirBuilder`.  Only the 2 I brought up previously, `ExecutionException` and `GeneralCommandLine` are `52.0`.

Full list (from `ElixirBuilder) with jar names for 2016.1:

```
// 52.0 in openapi.jar
// import com.intellij.execution.ExecutionException;
// 52.0 in openapi.jar
// import com.intellij.execution.configurations.GeneralCommandLine;
// 50.0 in util.jar
import com.intellij.execution.process.BaseOSProcessHandler;
// 50.0 in util.jar
import com.intellij.execution.process.ProcessAdapter;
// 50.0 in util.jar
import com.intellij.openapi.diagnostic.Logger;
// 50.0 in util.jar
import com.intellij.openapi.util.io.FileUtil;
// 49.0 in util.jar
import com.intellij.openapi.util.io.FileUtilRt;
// 50.0 in util.jar
import com.intellij.openapi.util.text.StringUtil;
// 50.0 in util.jar
import com.intellij.util.CommonProcessors;
// 49.0 in util.jar
import com.intellij.util.Function;
// 50.0 in util.jar
import com.intellij.util.containers.ContainerUtil;
// 49.0 in trove4j.jar
import gnu.trove.THashSet;
import org.elixir_lang.jps.mix.JpsMixConfigurationExtension;
import org.elixir_lang.jps.model.ElixirCompilerOptions;
import org.elixir_lang.jps.model.JpsElixirCompilerOptionsExtension;
import org.elixir_lang.jps.model.JpsElixirModuleType;
import org.elixir_lang.jps.model.JpsElixirSdkType;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
// 50.0 in jsp-builders.jar
import org.jetbrains.jps.builders.BuildOutputConsumer;
// 50.0 in jsp-builders.jar
import org.jetbrains.jps.builders.DirtyFilesHolder;
// 50.0 in jsp-builders.jar
import org.jetbrains.jps.builders.FileProcessor;
// 50.0 in jsp-builders.jar
import org.jetbrains.jps.incremental.CompileContext;
// 50.0 in jsp-builders.jar
import org.jetbrains.jps.incremental.ProjectBuildException;
// 50.0 in jsp-builders.jar
import org.jetbrains.jps.incremental.TargetBuilder;
// 50.0 in jsp-builders.jar
import org.jetbrains.jps.incremental.messages.BuildMessage;
// 50.0 in jsp-builders.jar
import org.jetbrains.jps.incremental.messages.CompilerMessage;
// 50.0 in jsp-builders.jar
import org.jetbrains.jps.incremental.resources.ResourcesBuilder;
// 50.0 in jsp-builders.jar
import org.jetbrains.jps.incremental.resources.StandardResourceBuilderEnabler;
// 50.0 in openapi.jar
import org.jetbrains.jps.model.JpsDummyElement;
// 50.0 in openapi.jar
import org.jetbrains.jps.model.JpsProject;
// 50.0 in openapi.jar
import org.jetbrains.jps.model.java.JavaSourceRootType;
// 50.0 in openapi.jar
import org.jetbrains.jps.model.java.JpsJavaExtensionService;
// 50.0 in openapi.jar
import org.jetbrains.jps.model.library.sdk.JpsSdk;
// 50.0 in openapi.jar
import org.jetbrains.jps.model.module.*;
```

**Comment by carymrobbins** [comment]
One thing that seems to work is to inspect the running Java version and only load your builders if they are supported.  Not pretty, but does seem to solve the problem.

[You can see my proposed fix here.](https://github.com/carymrobbins/intellij-haskforce/blob/b599c50db54a030df382b5c18a14697dce90f793/jps-plugin/src/com/haskforce/jps/HaskForceBuilder.java#L21-L25)

**Comment by ignatov** [comment]
Thanks.

**Comment by klesun** [comment]
tl;dr the thread, but the issue seems to be still relevant in 2021

disabling the "Erlang" plugin (`0.11.1108`) solves the issue for building java projects unrelated to erlang
